pkg/signature/algorithm_registry.go

```diff
@@ -0,0 +1,212 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package signature
+
+import (
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/ed25519"
+	"crypto/elliptic"
+	"crypto/rsa"
+	"fmt"
+
+	v1 "github.com/sigstore/protobuf-specs/gen/pb-go/common/v1"
+)
+
+// PublicKeyType represents the public key algorithm for a given signature algorithm.
+type PublicKeyType uint
+
+const (
+	RSA     PublicKeyType = iota // RSA public key
+	ECDSA                        // ECDSA public key
+	ED25519                      // ED25519 public key
+)
+
+// RSAKeySize represents the size of an RSA public key.
+type RSAKeySize int
+
+// AlgorithmDetails exposes relevant information for a given signature algorithm.
+type AlgorithmDetails interface {
+	// GetSignatureAlgorithm returns the algorithm registry.
+	GetSignatureAlgorithm() v1.KnownSignatureAlgorithm
+
+	// GetKeyType returns the public key algorithm for the given signature algorithm.
+	GetKeyType() PublicKeyType
+
+	// GetHashType returns the hash algorithm for a given signature algorithm.
+	GetHashType() crypto.Hash
+
+	// GetRSAKeySize returns the key size if the public key algorithm is RSA.
+	// Otherwise, an error is returned.
+	GetRSAKeySize() (RSAKeySize, error)
+
+	// GetECDSACurve returns the curve if the public key algorithm is ECDSA.
+	// Otherwise, an error is returned.
+	GetECDSACurve() (*elliptic.Curve, error)
+
+	checkKey(crypto.PublicKey) bool
+	checkHash(crypto.Hash) bool
+}
+
+type algorithmDetailsImpl struct {
+	knownAlgorithm v1.KnownSignatureAlgorithm
+	keyType        PublicKeyType
+	hashType       crypto.Hash
+	extraKeyParams interface{}
+	flagValue      string
+}
+
+func (a algorithmDetailsImpl) GetSignatureAlgorithm() v1.KnownSignatureAlgorithm {
+	return a.knownAlgorithm
+}
+
+func (a algorithmDetailsImpl) GetKeyType() PublicKeyType {
+	return a.keyType
+}
+
+func (a algorithmDetailsImpl) GetHashType() crypto.Hash {
+	return a.hashType
+}
+
+func (a algorithmDetailsImpl) GetRSAKeySize() (RSAKeySize, error) {
+	if a.keyType != RSA {
+		return 0, fmt.Errorf("unable to retrieve RSA key size for key type: %T", a.keyType)
+	}
+	rsaKeySize, ok := a.extraKeyParams.(RSAKeySize)
+	if !ok {
+		panic("unable to retrieve key size for RSA, malformed algorithm details?")
+	}
+	return rsaKeySize, nil
+}
+
+func (a algorithmDetailsImpl) GetECDSACurve() (*elliptic.Curve, error) {
+	if a.keyType != ECDSA {
+		return nil, fmt.Errorf("unable to retrieve ECDSA curve for key type: %T", a.keyType)
+	}
+	ecdsaCurve, ok := a.extraKeyParams.(elliptic.Curve)
+	if !ok {
+		panic("unable to retrieve curve for ECDSA, malformed algorithm details?")
```

Comment:
I'm using `panic` pretty freely here since it's all dependent on what's in `algorithmDetails` and therefore, there shouldn't be any way for the user to reach this.

Reply:
I prefer to avoid panics, just in case some refactor occurs and this ends up being user-controlled (contrived, but possible). Given the API returns an error already, I'd just return an error.

I see there's a few places below that don't return an error, we might just need to plumb errors through.

Reply:
Sure, that should be easy to do.

Reply:
[12edb0f](https://github.com/sigstore/sigstore/pull/1601/commits/12edb0f94732c78348507edbb297d9330536511b)